map/chart_generator.cpp

```diff
@@ -0,0 +1,215 @@
+    return;
+  }
+
+  uniformAltitudeDataM.resize(resultPointCount);
+  double const routeLenM = distanceDataM.back();
+  double const stepLenM = routeLenM / static_cast<double>(resultPointCount - 1);
+
+  auto const calculateAltitude = [&](double distFormStartM) {
+    if (distFormStartM <= distanceDataM.front())
+      return static_cast<double>(altitudeDataM.front());
+    if (distFormStartM >= distanceDataM.back())
+      return static_cast<double>(altitudeDataM.back());
+
+    auto const lowerIt = lower_bound(distanceDataM.cbegin(), distanceDataM.cend(), distFormStartM);
+    if (lowerIt == distanceDataM.cbegin())
+      return static_cast<double>(altitudeDataM.front());
+    if (lowerIt == distanceDataM.cend())
+      return static_cast<double>(altitudeDataM.back());
+
+    size_t const nextPointIdx = distance(distanceDataM.cbegin(), lowerIt);
+    CHECK_LESS(0, nextPointIdx, ());
+    size_t const prevPointIdx = nextPointIdx - 1;
+
+    if (distanceDataM[prevPointIdx] == distanceDataM[nextPointIdx])
+      return static_cast<double>(altitudeDataM[prevPointIdx]);
+
+    double const k = (altitudeDataM[nextPointIdx] - altitudeDataM[prevPointIdx]) /
+                     (distanceDataM[nextPointIdx] - distanceDataM[prevPointIdx]);
+    return static_cast<double>(altitudeDataM[prevPointIdx]) +
+           k * (distFormStartM - distanceDataM[prevPointIdx]);
+  };
+
+  uniformAltitudeDataM.resize(resultPointCount);
+  for (size_t i = 0; i < resultPointCount; ++i)
+    uniformAltitudeDataM[i] = calculateAltitude(static_cast<double>(i) * stepLenM);
+}
+
+void GenerateYAxisChartData(size_t height, double minMetersPerPxl,
+                            vector<double> const & altitudeDataM, vector<double> & yAxisDataPxl)
+{
+  uint32_t constexpr kHeightIndentPxl = 2.0;
+  yAxisDataPxl.clear();
+  uint32_t heightIndent = kHeightIndentPxl;
+  if (height <= 2 * kHeightIndentPxl)
+  {
+    LOG(LERROR, ("Chart height is less then 2 * kHeightIndentPxl (", 2 * kHeightIndentPxl, ")"));
+    heightIndent = 0;
+  }
+
+  auto const minMaxAltitudeIt = minmax_element(altitudeDataM.begin(), altitudeDataM.end());
+  double const minAltM = *minMaxAltitudeIt.first;
+  double const maxAltM = *minMaxAltitudeIt.second;
+  double const deltaAltM = maxAltM - minAltM;
+  uint32_t const drawHeightPxl = height - 2 * heightIndent;
+  double const meterPerPxl = max(minMetersPerPxl, deltaAltM / static_cast<double>(drawHeightPxl));
+  if (meterPerPxl == 0.0)
+  {
+    LOG(LERROR, ("meterPerPxl == 0.0"));
+    return;
+  }
+
+  size_t const altitudeDataSz = altitudeDataM.size();
+  yAxisDataPxl.resize(altitudeDataSz);
+  for (size_t i = 0; i < altitudeDataSz; ++i)
+    yAxisDataPxl[i] = height - heightIndent - (altitudeDataM[i] - minAltM) / meterPerPxl;
+}
+
+void GenerateChartByPoints(size_t width, size_t height, vector<m2::PointD> const & geometry,
+                           bool day, vector<uint8_t> & frameBuffer)
+{
+  frameBuffer.clear();
```

Comment:
Yes. But GenerateChartByPoints returns nothing. I guess it'll be considered as success. So there's an alternative: either returning bool or clearing |frameBuffer|.
What do you think?